aiida/backends/sqlalchemy/migrations/versions/d102e96ea943_inline_calc_attributes.py

```diff
@@ -0,0 +1,130 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Migrate attributes from old InlineCalculation that have been transformed into CalcFunctions:
+
+* OLD `sealed` and `function_name` remain the same
+* OLD `namespace` becomes `function_namespace`
+* OLD `first_line_source_code` becomes `function_starting_line_number`
+* OLD `source_file` is moved to repository and dropped from DB
+* OLD `source_code` is used to calculate `function_number_of_lines` and dropped from DB
+* NEW `exit_status` is initialized at 0
+* NEW `process_state` is initialized at "finished"
+* NEW `process_label` is set as `Legacy InlineCalculation`
+
+Note 1: old InlineCalculation can be identified by the old keywords, and after migration they will be
+identified by having `function_number_of_lines` but no `version`.
+
+Note 2: `exit_status` and `process_state` are initialized to 0 and "finished" respectively because the
+CalcFunctions affected are old InlineCalculation, which were only stored if they succeeded. You can check
+the relevant code here (line 60 has the execution call, it is stored after that):
+
+https://github.com/aiidateam/aiida-core/blob/v0.12.5/aiida/orm/implementation/django/calculation/inline.py#L60
+
+Revision ID: d102e96ea943
+Revises: 0edcdd5a30f0
+Create Date: 2021-02-09 16:25:16.106176
+
+"""
+# pylint: disable=invalid-name,no-member
+# yapf: disable
+from alembic import op
+from sqlalchemy.sql import text
+from aiida.backends.general.migrations import utils
+
+# revision identifiers, used by Alembic.
+revision = 'd102e96ea943'
+down_revision = '0edcdd5a30f0'
+branch_labels = None
+depends_on = None
+
+
+def upgrade():
+    """Creates the new attributes"""
+    conn = op.get_bind()
+
+    # Creates 'function_namespace' from 'namespace'
+    # Creates 'function_starting_line_number' from 'first_line_source_code'
+    # Creates 'exit_status', 'process_state' and 'process_label'
+    conn.execute(text(
+        """
+        UPDATE db_dbnode
+        SET attributes = attributes || '{
+            "exit_status": 0,
+            "process_state": "finished",
+            "process_label": "Legacy InlineCalculation"
+            }'::jsonb ||
+            jsonb_build_object(
+                'function_starting_line_number', attributes->'first_line_source_code',
+                'function_namespace', attributes->'namespace'
```

Comment:
Does this _need_ to be separate from the first dictionary? Couldn't this be merged? Or do you have to use `jsonb_build_object` here because you need to access existing `attributes`? If so, might be worth a comment

Reply:
Indeed, as you say, I need the `jsonb_build_object` since I am extracting the info from the attributes of the each entry. What I could do is use `jsonb_build_object` for all:

```SQL
        SET attributes = attributes || jsonb_build_object(
                'exit_status': 0,
                'process_state': 'finished',
                'process_label': 'Legacy InlineCalculation',
                'function_starting_line_number', attributes->'first_line_source_code',
                'function_namespace', attributes->'namespace'
                )
```

I did it this way because I thought it could be useful to have in the codebase as a reference the two ways of doing so (the one that is sufficient for fixed modifications, and the one you need if you are transforming data for each entry). What would you prefer, that I use the `jsonb_build_object` for all or that I leave both options and include a comment explaining?

Reply:
I think it is better to have it consistent and use the one method. I see what you were trying to accomplish but using both just makes the code more difficult to read

Reply:
So, I already changed it to use only `jsonb_build_object`, but do notice the horrible way of having to provide the dict as a list of `key1, val1, key2, val2, ...` instead of just providing a dict...:anguished: